components/ipfs/pin/ipfs_local_pin_service.cc

```diff
@@ -0,0 +1,284 @@
+                                        base::Unretained(this)));
+}
+
+void VerifyLocalPinJob::OnGetPinsResult(bool status,
+                                        absl::optional<GetPinsResult> result) {
+  if (status && result) {
+    DictionaryPrefUpdate update(prefs_service_, kIPFSPinnedCids);
+    base::Value::Dict& update_dict = update->GetDict();
+
+    bool verification_pased = true;
+    for (const auto& cid : cids_) {
+      base::Value::List* list = update_dict.FindList(cid);
+      if (!list) {
+        verification_pased = false;
+      } else {
+        if (result->find(cid) != result->end()) {
+          list->EraseValue(base::Value(prefix_));
+          list->Append(base::Value(prefix_));
+        } else {
+          verification_pased = false;
+          list->EraseValue(base::Value(prefix_));
+        }
+        if (list->empty()) {
+          update_dict.Remove(cid);
+        }
+      }
+    }
+    std::move(callback_).Run(verification_pased);
+  } else {
+    std::move(callback_).Run(absl::nullopt);
+  }
+}
+
+GcJob::GcJob(PrefService* prefs_service,
+             IpfsService* ipfs_service,
+             GcCallback callback)
+    : prefs_service_(prefs_service),
+      ipfs_service_(ipfs_service),
+      callback_(std::move(callback)) {}
+GcJob::~GcJob() {}
+
+void GcJob::Start() {
+  ipfs_service_->GetPins(
+      absl::nullopt, kRecursiveMode, true,
+      base::BindOnce(&GcJob::OnGetPinsResult, base::Unretained(this)));
+}
+
+void GcJob::OnGetPinsResult(bool status, absl::optional<GetPinsResult> result) {
+  std::vector<std::string> cids_to_delete;
+  if (status && result) {
+    const base::Value::Dict& dict = prefs_service_->GetDict(kIPFSPinnedCids);
+    for (const auto& pair : result.value()) {
+      const base::Value::List* list = dict.FindList(pair.first);
+      if (!list || list->empty()) {
+        cids_to_delete.push_back(pair.first);
+      }
+    }
+
+    if (!cids_to_delete.empty()) {
+      ipfs_service_->RemovePin(
+          cids_to_delete,
+          base::BindOnce(&GcJob::OnPinsRemovedResult, base::Unretained(this)));
+    } else {
+      std::move(callback_).Run(true);
+    }
+  } else {
+    std::move(callback_).Run(false);
+  }
+}
+
+void GcJob::OnPinsRemovedResult(bool status,
+                                absl::optional<RemovePinResult> result) {
+  if (status && result) {
+    std::move(callback_).Run(true);
+  } else {
+    std::move(callback_).Run(false);
+  }
+}
+
+IpfsLocalPinService::IpfsLocalPinService(PrefService* prefs_service,
+                                         IpfsService* ipfs_service)
+    : prefs_service_(prefs_service), ipfs_service_(ipfs_service) {
+  ipfs_base_pin_service_ =
+      std::make_unique<IpfsBasePinService>(prefs_service_, ipfs_service_);
+  base::SequencedTaskRunnerHandle::Get()->PostDelayedTask(
+      FROM_HERE,
+      base::BindOnce(&IpfsLocalPinService::AddGcTask, base::Unretained(this)),
+      base::Minutes(1));
```

Comment:
I think there is no guarantee that after a minute delay from start IpfsLocalPinService is always alive. We might get browser shutdown crash here.